engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java

```diff
@@ -134,13 +138,59 @@ Object doPanicSentinel(
     throw self;
   }
 
-  @Specialization
+  @Specialization(guards = "warnings.hasWarnings(self)")
   Object doWarning(
       VirtualFrame frame,
       State state,
       UnresolvedSymbol symbol,
-      WithWarnings self,
-      Object[] arguments) {
+      Object self,
+      Object[] arguments,
+      @Cached MethodResolverNode methodResolverNode,
+      @CachedLibrary(limit = "10") WarningsLibrary warnings) {
+    Function function =
+        symbol
+            .getScope()
+            .lookupMethodDefinition(
+                EnsoContext.get(this).getBuiltins().warning().getEigentype(), symbol.getName());
+    if (function != null) {
+      // Warning's builtin type methods are static meaning that they have a synthetic `self`
+      // parameter. However, the constructed `InvokeFunctionNode`  is missing that
+      // information and the function, if called with `arguments` will not be fully applied.
+      // Hence, the synthetic construction of a new `InvokeFunctionNode` with the updated schema
+      // and call including an additional, dummy, argument.
+      Object[] arguments1;
+      if (!function.getSchema().isFullyApplied()) {
+        arguments1 = new Object[function.getSchema().getArgumentsCount()];
+        System.arraycopy(arguments, 0, arguments1, 1, arguments.length);
+        arguments1[0] = arguments[0];
+      } else {
+        arguments1 = arguments;
+      }
+      if (warningFunctionNode == null) {
```

Comment:
`warningFunctionNode` shall also be accepted as `@Cached` argument - e.g. when the specialization is triggered, you'll get the node and you don't have to mangle with the AST manually.